File: app/src/main/java/com/iamagamedev/mybrandnewgame/gameObjects/Hero.java
package com.iamagamedev.mybrandnewgame.gameObjects;

import android.content.Context;

import com.iamagamedev.mybrandnewgame.Constants.CharConstants;
import com.iamagamedev.mybrandnewgame.Constants.ObjectNames;
import com.iamagamedev.mybrandnewgame.Constants.SpellNames;
import com.iamagamedev.mybrandnewgame.RectHitBox;
import com.iamagamedev.mybrandnewgame.Utils;
import com.iamagamedev.mybrandnewgame.gameObjects.spells.HeroSpell;
import com.iamagamedev.mybrandnewgame.gameObjects.spells.ShieldSpellObject;

import java.util.ArrayList;

/**
 * Created by Михан on 17.05.2017.
 */

public class Hero extends GameObject {

    private HeroSpell spellObject;

    private ArrayList<String> spellObjects;

    private final float MAX_VELOCITY = 2;
    private int pixelPerMetre;

    private boolean pressingRight, pressingLeft, pressingUp, pressingDown;

    public Hero(Context context, float worldStartX, float worldStartY,
                int pixelsPerMetre) {

        this.pixelPerMetre = pixelsPerMetre;
        spellObjects = new ArrayList<>();
        spellObjects.add(SpellNames.FIREBALL);

        final float HEALTH = 50;
        final float HEIGHT = 1;
        final float WIDTH = 1;

        setHeight(HEIGHT);
        setWidth(WIDTH);
        setHealth(HEALTH);

        setxVelocity(0);
        setyVelocity(0);
        setFacing(LEFT);

        setMoves(true);
        setActive(true);
        setVisible(true);
        setType(CharConstants.PLAYER);

        setBitmapName(ObjectNames.MAGE);
        setBadBitmapName(ObjectNames.MAGE_DEAD);

        final int ANIMATION_FPS = 16;
        final int ANIMATION_FRAME_COUNT = 4;

        setAnimFps(ANIMATION_FPS);
        setAnimFrameCount(ANIMATION_FRAME_COUNT);
        setAnimated(pixelsPerMetre, true);

        setWorldLocation(worldStartX, worldStartY, 1);
    }


    @Override
    public void update(long fps) {
        if (pressingRight) {
            this.setxVelocity(MAX_VELOCITY);
        } else if (pressingLeft) {
            this.setxVelocity(-MAX_VELOCITY);
        } else {
            this.setxVelocity(0);
        }
        if (pressingUp) {
            this.setyVelocity(-MAX_VELOCITY);
        } else if (pressingDown) {
            this.setyVelocity(MAX_VELOCITY);
        } else {
            this.setyVelocity(0);
        }

        if (this.getxVelocity() > 0) {
            setFacing(RIGHT);
        } else if (this.getxVelocity() < 0) {
            setFacing(LEFT);
        } else if (this.getyVelocity() > 0) {
            setFacing(DOWN);
        } else if (this.getyVelocity() < 0) {
            setFacing(UP);
        }

        this.move(fps);

        setRectHitBox();
        //updateShieldLocation();
    }

    public void setPressingDown(boolean pressingDown) {
        this.pressingDown = pressingDown;
    }

    public void setPressingUp(boolean pressingUp) {
        this.pressingUp = pressingUp;
    }

    public void setPressingLeft(boolean pressingLeft) {
        this.pressingLeft = pressingLeft;
    }

    public void setPressingRight(boolean pressingRight) {
        this.pressingRight = pressingRight;
    }

    public void fireSpell() {
        spellObject = HeroSpell.Companion.getInstance(this.getWorldLocation().x,
                this.getWorldLocation().y, CharConstants.SPELL, pixelPerMetre);
        Utils.fireSpell(spellObject, this, SpellNames.FIREBALL);
    }

    private void updateShieldLocation() {
        ShieldSpellObject shieldSpellObject = ShieldSpellObject.Companion.getInstance(
                getWorldLocation().x, getWorldLocation().y, CharConstants.SHIELD, pixelPerMetre);
        shieldSpellObject.setWorldLocation(getWorldLocation().x - 0.5f, getWorldLocation().y - 0.5f, 0);
    }
}

File: app/src/main/java/com/iamagamedev/mybrandnewgame/gameObjects/Hero.kt
package com.iamagamedev.mybrandnewgame.gameObjects

import android.content.Context
import com.iamagamedev.mybrandnewgame.constants.CharConstants
import com.iamagamedev.mybrandnewgame.constants.ObjectNames
import com.iamagamedev.mybrandnewgame.constants.SpellNames
import com.iamagamedev.mybrandnewgame.LevelManager
import com.iamagamedev.mybrandnewgame.Utils.fireSpell
import com.iamagamedev.mybrandnewgame.gameObjects.spells.HeroSpell
import com.iamagamedev.mybrandnewgame.gameObjects.spells.SpellObject
import java.util.*

/**
 * Created by Михан on 17.05.2017.
 */
class Hero(context: Context?, worldStartX: Float, worldStartY: Float,
           private val pixelPerMetre: Int) : GameObject() {
    private val spellObjects: ArrayList<String> = ArrayList()
    private val MAX_VELOCITY = 2f
    private var pressingRight = false
    private var pressingLeft = false
    private var pressingUp = false
    private var pressingDown = false
    override fun update(fps: Long) {
        when {
            pressingRight -> {
                setxVelocity(MAX_VELOCITY)
            }
            pressingLeft -> {
                setxVelocity(-MAX_VELOCITY)
            }
            else -> {
                setxVelocity(0f)
            }
        }
        when {
            pressingUp -> {
                setyVelocity(-MAX_VELOCITY)
            }
            pressingDown -> {
                setyVelocity(MAX_VELOCITY)
            }
            else -> {
                setyVelocity(0f)
            }
        }
        when {
            getxVelocity() > 0 -> {
                facing = RIGHT
            }
            getxVelocity() < 0 -> {
                facing = LEFT
            }
            getyVelocity() > 0 -> {
                facing = DOWN
            }
            getyVelocity() < 0 -> {
                facing = UP
            }
        }
        move(fps)
        setRectHitBox()
        //updateShieldLocation();
    }

    fun setPressingDown(pressingDown: Boolean) {
        this.pressingDown = pressingDown
    }

    fun setPressingUp(pressingUp: Boolean) {
        this.pressingUp = pressingUp
    }

    fun setPressingLeft(pressingLeft: Boolean) {
        this.pressingLeft = pressingLeft
    }

    fun setPressingRight(pressingRight: Boolean) {
        this.pressingRight = pressingRight
    }

    fun fireSpell() {
        if (LevelManager.spellObject === null) {
            LevelManager.spellObject = HeroSpell(this.worldLocation!!.x, this.worldLocation!!.y, CharConstants.SPELL, pixelPerMetre)
        }
        LevelManager.spellObject?.let { fireSpell(it, this@Hero, SpellNames.FIREBALL) }
    }

    private fun updateShieldLocation() {
        LevelManager.shieldObject!!.setWorldLocation(worldLocation!!.x - 0.5f, worldLocation!!.y - 0.5f, 0)
    }

    init {
        spellObjects.add(SpellNames.FIREBALL)
        val HEALTH = 50
        val HEIGHT = 1f
        val WIDTH = 1f
        height = HEIGHT
        width = WIDTH
        health = HEALTH
        setxVelocity(0f)
        setyVelocity(0f)
        facing = LEFT
        isMoves = true
        isActive = true
        isVisible = true
        type = CharConstants.PLAYER
        bitmapName = ObjectNames.MAGE
        badBitmapName = ObjectNames.MAGE_DEAD
        val ANIMATION_FPS = 16
        val ANIMATION_FRAME_COUNT = 4
        setAnimFps(ANIMATION_FPS)
        setAnimFrameCount(ANIMATION_FRAME_COUNT)
        setAnimated(pixelPerMetre, true)
        setWorldLocation(worldStartX, worldStartY, 1)
    }
}